framework/include/userobject/BulkMaterial.h

```diff
@@ -0,0 +1,242 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ElementUserObject.h"
+
+#include "libmesh/int_range.h"
+
+#include <tuple>
+#include <string>
+#include <vector>
+#include <utility>
+
+// type wrappers
+template <typename T>
+struct GatherMatProp
+{
+  typedef T type;
+  typedef const MaterialProperty<T> ref;
+
+  template <typename M>
+  static ref * getPointer(M & mpi, const MaterialPropertyName & name)
+  {
+    return &mpi.template getMaterialProperty<T>(name);
+  }
+};
+
+struct GatherVariable
+{
+  typedef Real type;
+  typedef const VariableValue ref;
+
+  template <typename C>
+  static ref * getPointer(const C & coupleable, const VariableName & name)
+  {
+    return &coupleable.coupledValue(name);
+  }
+};
+
+// tuple wrappers
+struct TupleStd
```

Comment:
This stuff seems really general and seems like it should be in a more general header? And especially something like `TupleStd` feels like it should be namespaced

Reply:
I namespaced it for now in `BulkMaterialUtils`. If we end up using the tuple wrappers elsewhere we can move them then.